File: cobalt/src/main/java/com/hartwig/hmftools/cobalt/targeted/TargetedRatioBuilder.java
package com.hartwig.hmftools.cobalt.targeted;

import static java.lang.Math.round;
import static java.lang.String.format;

import static com.hartwig.hmftools.cobalt.CobaltConfig.CB_LOGGER;
import static com.hartwig.hmftools.cobalt.CobaltUtils.replaceColumn;

import java.util.List;
import java.util.stream.Collectors;

import com.hartwig.hmftools.cobalt.ChromosomePositionCodec;
import com.hartwig.hmftools.cobalt.CobaltColumns;
import com.hartwig.hmftools.cobalt.CobaltConstants;
import com.hartwig.hmftools.cobalt.lowcov.LowCoverageRatioBuilder;
import com.hartwig.hmftools.cobalt.ratio.GcNormalizedRatioBuilder;
import com.hartwig.hmftools.cobalt.ratio.RatioBuilder;
import com.hartwig.hmftools.common.cobalt.ImmutableReadRatio;
import com.hartwig.hmftools.common.cobalt.ReadRatio;
import com.hartwig.hmftools.common.genome.position.GenomePosition;
import com.hartwig.hmftools.common.utils.Doubles;

import org.apache.commons.lang3.Validate;
import org.apache.logging.log4j.Level;
import org.jetbrains.annotations.Nullable;

import tech.tablesaw.api.DoubleColumn;
import tech.tablesaw.api.Table;
import tech.tablesaw.columns.numbers.NumberPredicates;

public class TargetedRatioBuilder implements RatioBuilder
{
    private Table mOnTargetRatios;
    private Table mOffTargetRatios;
    private Table mCombinedRatios;

    private final ChromosomePositionCodec mChromosomePosCodec;

    public TargetedRatioBuilder(final Table rawRatios, final Table targetRegionEnrichment,
            ChromosomePositionCodec chromosomePosCodec)
    {
        mOnTargetRatios = Table.create();
        mOffTargetRatios = Table.create();
        mCombinedRatios = Table.create();
        mChromosomePosCodec = chromosomePosCodec;

        Validate.isTrue(rawRatios.longColumn(CobaltColumns.ENCODED_CHROMOSOME_POS).isMissing().isEmpty());
        Validate.isTrue(targetRegionEnrichment.longColumn(CobaltColumns.ENCODED_CHROMOSOME_POS).isMissing().isEmpty());

        // merge in the targeted region columns
        Table ratiosWithTargetRegion = rawRatios.joinOn(CobaltColumns.ENCODED_CHROMOSOME_POS).inner(targetRegionEnrichment);

        // resort it, the join messes up with the ordering
        ratiosWithTargetRegion = ratiosWithTargetRegion.sortAscendingOn(CobaltColumns.ENCODED_CHROMOSOME_POS);

        CB_LOGGER.debug("ratios with targeted region: {}", ratiosWithTargetRegion);

        populateOnTargetRatios(ratiosWithTargetRegion, targetRegionEnrichment);
        // populateOffTargetRatios(ratiosWithTargetRegion, CobaltConstants.OFF_TARGET_WINDOW_SIZE, targetRegionEnrichment);
        CB_LOGGER.debug("{} on target GC ratios, {} off target GC ratios", mOnTargetRatios.rowCount(), mOffTargetRatios.rowCount());

        // populateCombinedRatios(mOnTargetRatios, mOffTargetRatios);
    }

    // we use on target ratios only for now
    @Override
    public Table ratios() { return mOnTargetRatios; }

    public Table onTargetRatios() { return mOnTargetRatios; }
    public Table offTargetRatios() { return mOffTargetRatios; }

    private void populateCombinedRatios(final Table ratios1, final Table ratios2)
    {
        mCombinedRatios = ratios1.append(ratios2);
    }

    private void populateOnTargetRatios(final Table rawRatios, final Table targetRelativeEnrichment)
    {
        // find all the ratios that are inside the target enriched regions
        // we filter out all the regions with 0 gc normalised ratios, as they do not actually
        // correctly reflect the amount of enrichment, and also very rare

        Table onTargetRatios = rawRatios.copy();
        double targetRegionGcRatioMedian = onTargetRatios.doubleColumn(CobaltColumns.RATIO).filter(NumberPredicates.isNonNegative).median();

        CB_LOGGER.printf(Level.INFO, "targeted mode GC ratio median: %.3f", targetRegionGcRatioMedian);

        // normalise the ratio by relative enrichment and targeted region median
        DoubleColumn onTargetRatioColumn = onTargetRatios.doubleColumn(CobaltColumns.RATIO)
                .divide(onTargetRatios.doubleColumn("relativeEnrichment"))
                .divide(targetRegionGcRatioMedian)
                .setName(CobaltColumns.RATIO);

        onTargetRatios.replaceColumn(onTargetRatioColumn);

        mOnTargetRatios = onTargetRatios;
    }

    // we create a pan window ratio by taking the median count of super windows that combine multiple windows
    private void populateOffTargetRatios(final Table ratiosWithTargetRegion, final int offTargetWindowSize,
            final Table targetRegions)
    {
        Table offTargetRatios = ratiosWithTargetRegion.where(
                ratiosWithTargetRegion.booleanColumn("offTarget").asSelection()
                        .and(ratiosWithTargetRegion.doubleColumn("ratio").isNonNegative())
                        .and(ratiosWithTargetRegion.doubleColumn("relativeEnrichment").isNotMissing()));

        // double median = offTargetRatios.doubleColumn("ratio").median();

        // normalise the ratio by relative enrichment
        replaceColumn(offTargetRatios, "ratio",
                offTargetRatios.doubleColumn("ratio")
                        .divide(offTargetRatios.doubleColumn("relativeEnrichment")));

        CB_LOGGER.info("off target after enrichment normalisation: \n{}", offTargetRatios);

        // next we do low coverage
        offTargetRatios = new LowCoverageRatioBuilder(offTargetRatios, 1000, mChromosomePosCodec).ratios();

        CB_LOGGER.info("off target after consolidation: \n{}", offTargetRatios);

        // apply gc normalisation
        GcNormalizedRatioBuilder gcNormalizedRatioBuilder = new GcNormalizedRatioBuilder(offTargetRatios, true);
        offTargetRatios = gcNormalizedRatioBuilder.ratios();

        CB_LOGGER.info("off target gc normalisation: \n{}", gcNormalizedRatioBuilder.gcMedianReadCountTable());
        CB_LOGGER.info("off target after gc normalisation: \n{}", offTargetRatios);

        mOffTargetRatios = offTargetRatios;

        // remove any with invalid ratios
        // mOffTargetRatios = offTargetRatios.where(offTargetRatios.doubleColumn(CobaltColumns.RATIO).)

        /*
        Window window = new Window(offTargetWindowSize);

        for(String chromosome : rawRatios.stringColumn("chromosome").unique())
        {
            int currentWindowStart = 1;

            List<ReadRatio> windowGcRatios = new ArrayList<>();

            // we need this to make sure we get consistent chromosome name (1 vs chr1)


            for(ReadRatio readRatio : rawRatios.get(chromosome))
            {
                // todo: make sure this is sorted

                int windowStart = window.start(readRatio.position());

                if(windowStart != currentWindowStart)
                {
                    if(currentWindowStart != -1)
                    {
                        ReadRatio unnormalizedRatio = unnormalizedOffTargetRatio(
                            offTargetWindowSize, chromosomeStr, currentWindowStart, windowGcRatios, targetRegions);

                        if(unnormalizedRatio != null)
                        {
                            unnormalizedRatios.put(chromosome, unnormalizedRatio);
                        }
                    }

                    currentWindowStart = windowStart;
                    windowGcRatios.clear();
                }

                if(readRatio.ratio() >= 0)
                    windowGcRatios.add(readRatio);
            }

            ReadRatio unnormalizedRatio = unnormalizedOffTargetRatio(
                offTargetWindowSize, chromosomeStr, currentWindowStart, windowGcRatios, targetRegions);

            if(unnormalizedRatio != null)
                unnormalizedRatios.put(chromosome, unnormalizedRatio);
        }

        // now we want to normalise all of those off target gc ratios by the median
        List<Double> values = new ArrayList<>();
        unnormalizedRatios.values().forEach(x -> values.add(x.ratio()));
        double median = Doubles.median(values);

        CB_LOGGER.debug("normalizing {} off target windows ratio by median: {}", unnormalizedRatios.size(), median);

        mOffTargetRatios.clear();

        //for ((key, value) in unnormalizedRatios.entries())
        for(Map.Entry<Chromosome,ReadRatio> entry : unnormalizedRatios.entries())
        {
            ReadRatio readRatio = entry.getValue();
            double normalizedRatio = readRatio.ratio() / median;

            mOffTargetRatios.put(entry.getKey(), ImmutableReadRatio.builder().from(readRatio).ratio(normalizedRatio).build());
        }
         */
    }

    @Nullable
    private static ReadRatio unnormalizedOffTargetRatio(
            int offTargetWindowSize, final String chromosome, int windowStart, final List<ReadRatio> windowGcRatios,
            final List<GenomePosition> targetRegions)
    {
        int minNumGcRatios = (int)round(offTargetWindowSize / CobaltConstants.WINDOW_SIZE * CobaltConstants.MIN_OFF_TARGET_WINDOW_RATIO);

        int windowEnd = windowStart + offTargetWindowSize - 1;

        // the window position is the middle
        int windowMid = windowStart + offTargetWindowSize / 2;

        // check for targeted regions, we want to remove them
        for (GenomePosition targetRegion : targetRegions)
        {
            if (targetRegion.chromosome().equals(chromosome) && windowStart <= targetRegion.position() && windowEnd > targetRegion.position())
            {
                // this window contains a target region
                int removeStart = targetRegion.position() - 2 * CobaltConstants.WINDOW_SIZE;
                int removeEnd = targetRegion.position() + 2 * CobaltConstants.WINDOW_SIZE;
                CB_LOGGER.trace("off target window: {}:{} ({} - {}), contains target region",
                    chromosome, windowMid, windowStart, windowEnd);

                windowGcRatios.removeIf(o -> o.position() >= removeStart && o.position() <= removeEnd);
            }
        }

        if (windowGcRatios.size() < minNumGcRatios)
        {
            // if we don't have enough sub windows with valid values then we skip this
            CB_LOGGER.trace( "off target window: {}:{} ({} - {}), not enough sub window",
                    chromosome, windowMid, windowStart, windowEnd);
            return null;
        }

        double median = Doubles.median(windowGcRatios.stream().map(ReadRatio::ratio).collect(Collectors.toList()));

        CB_LOGGER.debug("off target window: {}:{} ({} - {}), num sub windows: {}, median: {}",
                chromosome, windowMid, windowStart, windowEnd, windowGcRatios.size(), format("%.4f", median));

        return !Double.isNaN(median) ? ImmutableReadRatio.builder().chromosome(chromosome).position(windowMid).ratio(median).build() : null;
    }
}

File: cobalt/src/main/java/com/hartwig/hmftools/cobalt/targeted/TargetedRatioBuilder.kt
package com.hartwig.hmftools.cobalt.targeted

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.ListMultimap
import com.google.common.collect.Sets
import com.hartwig.hmftools.cobalt.Chromosome
import com.hartwig.hmftools.cobalt.CobaltConstants
import com.hartwig.hmftools.cobalt.ratio.RatioBuilder
import com.hartwig.hmftools.common.cobalt.ImmutableReadRatio
import com.hartwig.hmftools.common.cobalt.ReadRatio
import com.hartwig.hmftools.common.genome.position.GenomePosition
import com.hartwig.hmftools.common.genome.window.Window
import com.hartwig.hmftools.common.utils.Doubles
import org.apache.logging.log4j.Level
import org.apache.logging.log4j.LogManager
import kotlin.math.roundToInt

class TargetedRatioBuilder(
    targetRegions: List<GenomePosition>,
    targetEnrichmentRatios: Map<GenomePosition, Double>,
    rawRatios: ListMultimap<Chromosome, ReadRatio>) : RatioBuilder
{
    private val mOnTargetRatios = ArrayListMultimap.create<Chromosome, ReadRatio>()
    private val mOffTargetRatios = ArrayListMultimap.create<Chromosome, ReadRatio>()
    private val mCombinedRatios = ArrayListMultimap.create<Chromosome, ReadRatio>()

    // we use on target ratios only for now
    override fun ratios(): ArrayListMultimap<Chromosome, ReadRatio>
    {
        return mOnTargetRatios
    }

    fun onTargetRatios(): ArrayListMultimap<Chromosome, ReadRatio>
    {
        return mOnTargetRatios
    }

    fun offTargetRatios(): ArrayListMultimap<Chromosome, ReadRatio>
    {
        return mOffTargetRatios
    }

    init
    {
        // calculate median for sample
        populateOnTargetRatios(targetEnrichmentRatios, rawRatios)
        populateOffTargetRatios(rawRatios, CobaltConstants.OFF_TARGET_WINDOW_SIZE, targetRegions)
        sLogger.info("{} on target GC ratios, {} off target GC ratios", mOnTargetRatios.size(), mOffTargetRatios.size())
        populateCombinedRatios(mOnTargetRatios, mOffTargetRatios)
    }

    private fun populateCombinedRatios(ratios1: ArrayListMultimap<Chromosome, ReadRatio>,
                                       ratios2: ArrayListMultimap<Chromosome, ReadRatio>)
    {
        mCombinedRatios.clear()
        val chromosomes = Sets.newIdentityHashSet<Chromosome>()
        chromosomes.addAll(ratios1.keySet())
        chromosomes.addAll(ratios2.keySet())
        for (chromosome in chromosomes)
        {
            var i1 = 0
            var i2 = 0
            val list1 = ratios1[chromosome]
            val list2 = ratios2[chromosome]
            while (i1 < list1.size && i2 < list2.size)
            {
                val readRatio1 = list1[i1]
                val readRatio2 = list2[i2]
                if (readRatio1.position() < readRatio2.position())
                {
                    mCombinedRatios.put(chromosome, readRatio1)
                    ++i1
                }
                else if (readRatio1.position() > readRatio2.position())
                {
                    mCombinedRatios.put(chromosome, readRatio2)
                    ++i2
                }
                else
                {
                    // shouldn't be here
                    sLogger.error(
                        "on target and off target ratio at same position: {}:{}",
                        readRatio1.chromosome(), readRatio1.position())
                    throw RuntimeException("on target and off target ratio at same position")
                }
            }

            // put the rest in
            while (i1 < list1.size)
            {
                mCombinedRatios.put(chromosome, list1[i1])
                ++i1
            }
            while (i2 < list2.size)
            {
                mCombinedRatios.put(chromosome, list2[i2])
                ++i2
            }
        }
    }

    private fun populateOnTargetRatios(
        targetRelativeEnrichment: Map<GenomePosition, Double>,
        rawRatios: ListMultimap<Chromosome, ReadRatio>)
    {
        // find all the ratios that are inside the target enriched regions
        // we filter out all the regions with 0 gc normalised ratios, as they do not actually
        // correctly reflect the amount of enrichment, and also very rare
        val targetRegionsGcRatios: List<Double> = rawRatios.entries()
            .filter { (_, readRatio): Map.Entry<Chromosome, ReadRatio> ->
                readRatio.ratio() > 0.0 && targetRelativeEnrichment.containsKey(readRatio)
            }
            .map { (_, value): Map.Entry<Chromosome, ReadRatio> -> value.ratio() }

        var targetRegionGcRatioMedian = 1.0

        if (targetRegionsGcRatios.isNotEmpty())
        {
            targetRegionGcRatioMedian = Doubles.median(targetRegionsGcRatios)
        }
        else
        {
            sLogger.warn("target region gc ratios is empty")
        }

        sLogger.printf(Level.INFO, "targeted mode GC ratio median: %.3f", targetRegionGcRatioMedian)
        mOnTargetRatios.clear()
        for ((key, value) in rawRatios.entries())
        {
            val relativeEnrichment = targetRelativeEnrichment[value] ?: continue
            val enrichmentAdjRatio = value.ratio() / targetRegionGcRatioMedian / relativeEnrichment
            sLogger.debug(
                "{}:{} relative enrichment: {}, on target ratio: {}",
                value.chromosome(),
                value.position(),
                relativeEnrichment,
                enrichmentAdjRatio)
            mOnTargetRatios.put(key, ImmutableReadRatio.builder().from(value).ratio(enrichmentAdjRatio).build())
        }
    }

    // we create a pan window ratio by taking the median count of super windows that combine multiple windows
    private fun populateOffTargetRatios(
        rawRatios: ListMultimap<Chromosome, ReadRatio>, offTargetWindowSize: Int,
        targetRegions: List<GenomePosition>)
    {
        val unnormalizedRatios = ArrayListMultimap.create<Chromosome, ReadRatio>()
        val window = Window(offTargetWindowSize)
        for (chromosome in rawRatios.keySet())
        {
            var currentWindowStart: Int = -1
            val windowGcRatios = ArrayList<Double>()

            // we need this to make sure we get consistent chromosome name (1 vs chr1)
            val chromosomeStr: String = rawRatios[chromosome][0].chromosome()
            for (readRatio in rawRatios[chromosome])
            {
                // todo: make sure this is sorted
                val windowStart: Int = window.start(readRatio.position())
                if (windowStart != currentWindowStart)
                {
                    if (currentWindowStart != -1)
                    {
                        val unnormalizedRatio = unnormalizedOffTargetRatio(
                            offTargetWindowSize,
                            chromosomeStr,
                            currentWindowStart,
                            windowGcRatios,
                            targetRegions)
                        if (unnormalizedRatio != null)
                        {
                            unnormalizedRatios.put(chromosome, unnormalizedRatio)
                        }
                    }
                    currentWindowStart = windowStart
                    windowGcRatios.clear()
                }
                if (readRatio.ratio() >= 0.0)
                {
                    windowGcRatios.add(readRatio.ratio())
                }
            }
            val unnormalizedRatio = unnormalizedOffTargetRatio(
                offTargetWindowSize,
                chromosomeStr,
                currentWindowStart,
                windowGcRatios,
                targetRegions)
            if (unnormalizedRatio != null)
            {
                unnormalizedRatios.put(chromosome, unnormalizedRatio)
            }
        }

        // now we want to normalise all of those off target gc ratios by the median
        val median: Double = Doubles.median(unnormalizedRatios.values().map({ obj: ReadRatio -> obj.ratio() }))
        sLogger.debug("normalizing {} off target windows ratio by median: {}", unnormalizedRatios.size(), median)
        mOffTargetRatios.clear()
        for ((key, value) in unnormalizedRatios.entries())
        {
            val normalizedRatio = value.ratio() / median
            mOffTargetRatios.put(
                key,
                ImmutableReadRatio.builder().from(value).ratio(normalizedRatio).build())
        }
    }

    companion object
    {
        private val sLogger = LogManager.getLogger(TargetedRatioBuilder::class.java)
        private fun unnormalizedOffTargetRatio(offTargetWindowSize: Int,
                                               chromosome: String,
                                               windowStart: Int,
                                               windowGcRatios: List<Double>,
                                               targetRegions: List<GenomePosition>): ReadRatio?
        {
            val minNumGcRatios =
                (offTargetWindowSize.toDouble() / CobaltConstants.WINDOW_SIZE * CobaltConstants.MIN_OFF_TARGET_WINDOW_RATIO).roundToInt()
            if (windowGcRatios.size < minNumGcRatios)
            {
                // if we don't have enough sub windows with valid values then we skip this
                return null
            }
            val windowEnd = windowStart + offTargetWindowSize - 1

            // check that this window does not contain any target regions
            for (targetRegion in targetRegions)
            {
                if (windowStart <= targetRegion.position() && windowEnd > targetRegion.position())
                {
                    // this window contains a target region
                    return null
                }
            }
            val median = Doubles.median(windowGcRatios)

            // the window position is the middle
            val windowMid = windowStart + offTargetWindowSize / 2
            sLogger.debug(
                "off target window: {}:{} ({} - {}), num sub windows: {}, median: {}",
                chromosome, windowMid,
                windowStart, windowEnd, windowGcRatios.size, median)
            if (!median.isNaN())
            {
                return ImmutableReadRatio.builder().chromosome(chromosome).position(windowMid).ratio(median).build()
            }
            return null
        }
    }
}